jax/numpy/linalg.py

```diff
@@ -148,12 +148,88 @@ def _slogdet_jvp(primals, tangents):
   return (sign, ans), (sign_dot, ans_dot)
 
 
+def _cofactor_solve(a, b):
+  """Equivalent to det(a)*solve(a, b) for nonsingular mat.
+
+  Intermediate function used for jvp and vjp of det.
+  This function borrows heavily from jax.numpy.linalg.solve and
+  jax.numpy.linalg.slogdet to compute the gradient of the determinant
+  in a way that is well defined even for low rank matrices.
+  * assumes a is at least rank n-1
+  * assumes u_{nn} is the element set to 0 in rank n-1 cases.
+
+  Args:
+    a: A square matrix or batch of matrices, possibly singular.
+    b: A matrix, or batch of matrices of the same dimension as a.
+
+  Returns:
+    det(a) and cofactor(a)^T*b, aka adjugate(a)*b
+  """
+  a = _promote_arg_dtypes(np.asarray(a))
+  b = _promote_arg_dtypes(np.asarray(b))
+  a_shape = np.shape(a)
+  b_shape = np.shape(b)
+  a_ndims = len(a_shape)
+  b_ndims = len(b_shape)
+  if not (a_ndims >= 2 and a_shape[-1] == a_shape[-2]
+    and b_shape[-2:] == a_shape[-2:]):
+    msg = ("The arguments to _cofactor_solve must have shapes "
+           "a=[..., m, m] and b=[..., m, m]; got a={} and b={}")
+    raise ValueError(msg.format(a_shape, b_shape))
+  if a_shape[-1] == 1:
+    return a[0, 0], b
+  # lu contains u in the upper triangular matrix and l in the strict lower
+  # triangular matrix.
+  # The diagonal of l is set to ones without loss of generality.
+  lu, pivots = lax_linalg.lu(a)
+  dtype = lax.dtype(a)
+  batch_dims = lax.broadcast_shapes(lu.shape[:-2], b.shape[:-2])
+  x = np.broadcast_to(b, batch_dims + b.shape[-2:])
+  lu = np.broadcast_to(lu, batch_dims + lu.shape[-2:])
```

Comment:
You probably already have this working, but I'll note that you could probably simplify this considerably if you make use of `jnp.vectorize()` to adding batch dimensions with `vmap` instead of writing them by hand.

Reply:
This is taken pretty much line-by-line from an older version of `jnp.linalg.solve`. Is that what  is used there now?

Reply:
Actually you'd previously mentioned some things about changes to `solve` related to custom gradients, maybe worth taking a look at the new version to see if we can port some of those changes over. This also still doesn't work with gradients-of-gradients, so perhaps some of the tricks from `custom_solve` can be useful there.

Reply:
From a quick look - it seems pretty complicated! I'll keep this in mind for the future - but as you said, this is working now.

Reply:
Right, we're using `vectorize` inside solve now, in particular for the `_lu_solve_core` helper function:
https://github.com/google/jax/blob/d2653a1e8a29dda3f2df40098076c83978a0098f/jax/lax_linalg.py#L670

I'll have to think a little bit more about the custom_linear_solve thing....

Reply:
I'm not familiar with `jnp.vectorize`. Is it really just a matter of wrapping the function call with that decorator and removing the 3 lines you highlighted?

Reply:
That's roughly my understanding. We can iterate on that in follow-up PRs though!